tests/src/tests/test_utils.go

```diff
@@ -0,0 +1,324 @@
+	if !ok {
+		return tmpDir
+	}
+	testFile := filepath.Base(fn)
+	testFileExtension := filepath.Ext(testFile)
+	subdir := testFile[0 : len(testFile)-len(testFileExtension)]
+	return CreateTempDir(subdir)
+}
+
+func CreateTempDir(subdir string) string {
+	tempDir := tmpDir + subdir + "/"
+	if _, err := os.Stat(tempDir); os.IsNotExist(err) {
+		os.Mkdir(tempDir, 0755)
+	}
+	return tempDir
+}
+
+func GetDevFileName() string {
+	pc, fn, _, ok := runtime.Caller(1)
+	if !ok {
+		return tmpDir + "DefaultDevfile"
+	}
+
+	testFile := filepath.Base(fn)
+	testFileExtension := filepath.Ext(testFile)
+	subdir := testFile[0 : len(testFile)-len(testFileExtension)]
+	destDir := CreateTempDir(subdir)
+	callerName := runtime.FuncForPC(pc).Name()
+
+	pos1 := strings.LastIndex(callerName, "/tests.") + len("/tests.")
+	devfileName := destDir + callerName[pos1:len(callerName)] + ".yaml"
+
+	LogMessage(fmt.Sprintf("GetDevFileName : %s", devfileName))
+
+	return devfileName
+}
+
+func AddSuffixToFileName(fileName string, suffix string) string {
+	pos1 := strings.LastIndex(fileName, ".yaml")
+	newFileName := fileName[0:pos1] + suffix + ".yaml"
+	LogMessage(fmt.Sprintf("Add suffix %s to fileName %s : %s", suffix, fileName, newFileName))
+	return newFileName
+}
+
+func LogMessage(message string) string {
+	testLogger.Println(message)
+	return message
+}
+
+type TestDevfile struct {
+	SchemaDevFile   schema.Devfile
+	FileName        string
+	ParsedSchemaObj parser.DevfileObj
+	SchemaParsed    bool
+}
+
+var StringCount int = 0
+
+var RndSeed int64 = time.Now().UnixNano()
+
+func GetRandomUniqueString(n int, lower bool) string {
+	StringCount++
+	return fmt.Sprintf("%s%04d", GetRandomString(n, lower), StringCount)
+}
+
+func setRandSeed() {
+	RndSeed++
+	rand.Seed(RndSeed)
+}
+
+const schemaBytes = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
+
+func GetRandomString(n int, lower bool) string {
+	setRandSeed()
+	b := make([]byte, n)
+	for i := range b {
+		b[i] = schemaBytes[rand.Intn(len(schemaBytes)-1)]
+	}
+	randomString := string(b)
+	if lower {
+		randomString = strings.ToLower(randomString)
+	}
+	return randomString
+}
+
+func GetGroupKinds() []schema.CommandGroupKind {
+	setRandSeed()
+	return []schema.CommandGroupKind{schema.BuildCommandGroupKind, schema.RunCommandGroupKind, schema.TestCommandGroupKind, schema.DebugCommandGroupKind}
+}
+
+func GetRandomGroupKind() schema.CommandGroupKind {
+	setRandSeed()
+	return GetGroupKinds()[rand.Intn(len(GetGroupKinds()))]
+}
+
+func GetBinaryDecision() bool {
+	return GetRandomDecision(1, 1)
+}
+
+func GetRandomDecision(success int, failure int) bool {
+	setRandSeed()
+	return rand.Intn(success+failure) > failure-1
+}
+
+func GetRandomNumber(max int) int {
+	setRandSeed()
+	return rand.Intn(max) + 1
+}
+
+func GetDevfile(fileName string) TestDevfile {
+	testDevfile := TestDevfile{}
+	testDevfile.SchemaDevFile = schema.Devfile{}
+	testDevfile.FileName = fileName
+	testDevfile.SchemaDevFile.SchemaVersion = "2.0.0"
```

Comment:
Just want to point out that I have seen this schema version is hard coded to `2.0.0`, and the tests are relying on devfile structs defined in api repo(schema). However, the parser repo is going to support multiple schema versions, not only V2, but also future versions. 
I know this is the first PR (part1) for the test, we might also want to start considering how the test structure is going to be for future schema versions, i.e. which parts can be commonly used, and which parts are version specific.

Reply:
Yes, that is on my to do list, not sure yet how to best do it.

Reply:
I honestly think we should do that before merging this PR. Once this merged, we will keep adding tests and before we know it, it will be a lot of updates and  refactoring to accommodate future version schema changes.